Add `binding` annotation to record an action path not resolved

upgle opened the pull request:
<!--- Provide a concise summary of your changes in the Title -->

## Description
<!--- Provide a detailed description of your changes. -->
<!--- Include details of what problem you are solving and how your changes are tested. -->

If the user invokes an action in a package binding, they can't know a package binding path from which action is invoked. because the current activation structure only records the resolved action path.

for example, suppose we have the following entities:

- `package` : **pkg1**
   - `action` : pkg1/**action1**
- `package` **pkg2** bind **pkg1**
- `package` **pkg3** bind **pkg1**

If user invokes the `pkg2/action1` action and the `pkg3/action1` action, all path annotations of invoked action in binding packages are saved as **namespace/pkg1/action1**.

```json
    "annotations": [
        {
            "key": "path",
            "value": "namespace/pkg1/action1"
        },
     ]
```

## Suggestion

So i added an annotation to distinguish between resolved action path and invoked action path and opend this PR.

```json
{
    "activationId": "f4250bb2721d4e4ca50bb2721d2e4cdd",
    "annotations": [
        {
            "key": "path",
            "value": "namespace/pkg1/action1"
        },
        {
            "key": "originPath",
            "value": "namespace/pkg2/action1"
        }
    ]
}
```



## Related issue and scope
<!--- Please include a link to a related issue if there is one. -->
- [ ] I opened an issue to propose and discuss this change (#????)

## My changes affect the following components
<!--- Select below all system components are affected by your change. -->
<!--- Enter an `x` in all applicable boxes. -->
- [ ] API
- [x] Controller
- [ ] Message Bus (e.g., Kafka)
- [ ] Loadbalancer
- [ ] Invoker
- [ ] Intrinsic actions (e.g., sequences, conductors)
- [ ] Data stores (e.g., CouchDB)
- [ ] Tests
- [ ] Deployment
- [ ] CLI
- [ ] General tooling
- [ ] Documentation

## Types of changes
<!--- What types of changes does your code introduce? Use `x` in all the boxes that apply: -->
- [ ] Bug fix (generally a non-breaking change which closes an issue).
- [x] Enhancement or new feature (adds new functionality).
- [ ] Breaking change (a bug fix or enhancement which changes existing behavior).

## Checklist:
<!--- Please review the points below which help you make sure you've covered all aspects of the change you're making. -->

- [x] I signed an [Apache CLA](https://github.com/apache/incubator-openwhisk/blob/master/CONTRIBUTING.md).
- [x] I reviewed the [style guides](https://github.com/apache/incubator-openwhisk/wiki/Contributing:-Git-guidelines#code-readiness) and followed the recommendations (Travis CI will check :).
- [x] I added tests to cover my changes.
- [ ] My changes require further changes to the documentation.
- [x] I updated the documentation where necessary.

rabbah commented:
I wonder if instead of modifying the Action type, we can add an alias to `FullyQualifiedName`. This would mean - if it works out - that you don't have to change the activation message and the name of the action and its alias (i.e., package and binding) are coupled more tightly. 

I would use `binding` in the annotation and leave the annotation as is otherwise (shows the package name).

This is a suggestion for discussion.

upgle commented:
As you mentioned, it would be better to add an alias to `FullyQualifiedName`.  I'll make a commit soon. 👍

upgle commented:
@rabbah 
I added `binding` field to `FullyQualifiedName` to not change the activation message, and renamed annotation key from `originPath` to `binding` and this annotation contains only entityPath of a package binding.

```json
{
    "activationId": "f4250bb2721d4e4ca50bb2721d2e4cdd",
    "annotations": [
        {
            "key": "binding",
            "value": "whisk.system/p2"
        },
        {
            "key": "path",
            "value": "guest/p1/hello"
        },
    ]
}
```

style95 commented:
@rabbah 
Is it possible to include this change? or any other direction is being considered?
In any case, I hope this issue is addressed so that we can resolve many binding related issues.

rabbah commented:
Thanks for the nudge. Will get on it.

upgle commented:
conflict is resolved

rabbah commented on common/scala/src/main/scala/org/apache/openwhisk/core/entity/WhiskAction.scala:
```diff
@@ -303,6 +327,10 @@ case class ExecutableWhiskActionMetaData(namespace: EntityPath,
     WhiskActionMetaData(namespace, name, exec, parameters, limits, version, publish, annotations)
       .revision[WhiskActionMetaData](rev)
+
+  def bindingFullyQualifiedName =
```
i suggest adding a type decorator on the method for the return value and a comment, explaining that the result is Some fully qualified name only if there's a binding, else None.

rabbah commented on common/scala/src/main/scala/org/apache/openwhisk/core/entity/WhiskAction.scala:
```diff
@@ -254,7 +274,8 @@ case class WhiskActionMetaData(namespace: EntityPath,
  * @param publish true to share the action or false otherwise
- * @param annotation the set of annotations to attribute to the action
+ * @param annotations the set of annotations to attribute to the action
+ * @param binding the path of the package binding
```
-> `@param binding the path of the package binding if any`

rabbah commented on common/scala/src/main/scala/org/apache/openwhisk/core/entity/WhiskActivation.scala:
```diff
@@ -137,6 +137,7 @@ object WhiskActivation
 
   /** Some field names for annotations */
   val pathAnnotation = "path"
+  val bindingAnnotation = "binding"
```
will check down below but this raises the opportunity to have the name in the activation be that of the binding, and the annotation is that of the package.

upgle replied:
I'm sorry, I don't understand this. 

In the example below, I think the name `triple` in the activation can not be the name of the binding `whisk.system/p2`

- there is an action `whisk.system/pkg1/triple`
- and the `whisk.system/p2` is binding the `whisk.system/pkg1`
- and i invoked an action `whisk.system/p2/triple`

```json
{
    "namespace": "whisk.system",
    "name": "triple",
    "version": "0.0.1",
    "subject": "whisk.system",
    "activationId": "b86606b16b8a4371a606b16b8a437182",
    "start": 1552900408086,
    "end": 1552900408090,
    "duration": 4,
    "annotations": [
        {
            "key": "binding",
            "value": "whisk.system/p2"
        },
        {
            "key": "path",
            "value": "whisk.system/pkg1/triple"
        },
...
```

style95 commented:
Any update on this PR?

upgle commented:
Sorry. I should resolve a conflict. I can handle today.

upgle commented:
@rabbah  A conflict is resolved. could you please review again?

c.c. @style95

style95 commented on core/controller/src/main/scala/org/apache/openwhisk/core/controller/actions/SequenceActions.scala:
```diff
@@ -312,13 +317,15 @@ protected[actions] trait SequenceActions {
   private def invokeNextAction(
     user: Identity,
     futureAction: Future[WhiskActionMetaData],
+    originAction: FullyQualifiedEntityName,
```
Is this being used in the call tree?

upgle replied:
It seems not using in the call tree. I'll fix it.

style95 commented on core/controller/src/main/scala/org/apache/openwhisk/core/controller/Actions.scala:
```diff
@@ -595,38 +590,6 @@ trait WhiskActionsApi extends WhiskCollectionAPI with PostActionActivation with
-   * Once the package is resolved, the operation is dispatched to the action in the package
-   * namespace.
-   */
-  private def mergeActionWithPackageAndDispatch(method: HttpMethod,
```
@upgle Could you elaborate the reason to remove this method a little bit?
It would help people who are not familiar with this change like me to get the intention more easily.

upgle replied:
If an action is resolved before it's dispatched, there is no way to know its original name in the activate method because it's already resolved.
So, I modified it to resolve inside the activate method after it has been dispatched.

```scala
override def activate(user: Identity, entityName: FullyQualifiedEntityName, env: Option[Parameters])(
    implicit transid: TransactionId) = {	    implicit transid: TransactionId)
```

upgle commented on core/controller/src/main/scala/org/apache/openwhisk/core/controller/Actions.scala:
```diff
@@ -253,7 +247,7 @@ trait WhiskActionsApi extends WhiskCollectionAPI with PostActionActivation with
       'timeout.as[FiniteDuration] ? WhiskActionsApi.maxWaitForBlockingActivation) { (blocking, result, waitOverride) =>
       entity(as[Option[JsObject]]) { payload =>
-        getEntity(WhiskActionMetaData.get(entityStore, entityName.toDocId), Some {
+        getEntity(WhiskActionMetaData.resolveActionAndMergeParameters(entityStore, entityName), Some {
```
@style95 the action is resolved here after it's dispatched.

style95 replied:
ok, this is the similar way which is taken for web action.

style95 commented:
@rabbah Do you have any comments on this?

style95 commented:
I would merge this in 72 hours if there is no objection based on silent consent.
